src/Microsoft.NetCore.Analyzers/Core/Security/UseXmlReaderForSchemaRead.cs

```diff
@@ -0,0 +1,112 @@
+            typeof(SystemSecurityCryptographyResources));
+        private static readonly LocalizableString s_Message = new LocalizableResourceString(
+            nameof(SystemSecurityCryptographyResources.UseXmlReaderForSchemaReadMessage),
+            SystemSecurityCryptographyResources.ResourceManager,
+            typeof(SystemSecurityCryptographyResources));
+        private static readonly LocalizableString s_Description = new LocalizableResourceString(
+            nameof(SystemSecurityCryptographyResources.UseXmlReaderForSchemaReadDescription),
+            SystemSecurityCryptographyResources.ResourceManager,
+            typeof(SystemSecurityCryptographyResources));
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+                DiagnosticId,
+                s_Title,
+                s_Message,
+                DiagnosticCategory.Security,
+                DiagnosticHelpers.DefaultDiagnosticSeverity,
+                isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                description: s_Description,
+                helpLinkUri: null,
+                customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(compilationStartAnalysisContext =>
+            {
+                var compilation = compilationStartAnalysisContext.Compilation;
+                var dataSetTypeSymbol = compilation.GetTypeByMetadataName(WellKnownTypeNames.SystemDataDataSet);
+
+                if (dataSetTypeSymbol == null)
+                {
+                    return;
+                }
+
+                var xmlReaderTypeSymbol = compilation.GetTypeByMetadataName(WellKnownTypeNames.SystemXmlXmlReader);
+
+                compilationStartAnalysisContext.RegisterOperationAction(operationAnalysisContext =>
+                {
+                    var invocationOperation = (IInvocationOperation)operationAnalysisContext.Operation;
+                    var methodSymbol = invocationOperation.TargetMethod;
+                    var methodName = methodSymbol.Name;
+
+                    if (methodName.StartsWith("ReadXml", StringComparison.Ordinal) &&
+                        IsOverrides(methodSymbol, dataSetTypeSymbol))
+                    {
+                        if (xmlReaderTypeSymbol != null &&
+                            methodSymbol.Parameters.Length > 0 &&
+                            methodSymbol.Parameters[0].Type.Equals(xmlReaderTypeSymbol))
```

Comment:
Would it be worth checking for any parameter being xmlReaderTypeSymbol, just in case one day DataSet's ReadXml*() methods change?

Reply:
I checked the other semmle rule UseXmlReaderFor***.ql, they all check the first parameter. Maybe it's a convention, I think if there's a new API, they are gonna put XmlReader at the first position, too. If there's a change, we can change later. =)